10535 -  \yii\db\Expression with complex expression (aggregate functions and etc) for Order By clause

andrewnester opened the pull request:
Added unit test for #10535

andrewnester committed:
Added test to prove orderBy accepts \yii\db\Expression #10535

cebe commented on tests/framework/db/QueryTest.php:
```diff
@@ -166,6 +167,9 @@ public function testOrder()
         $query->addOrderBy('age ASC, company DESC');
         $this->assertEquals(['team' => SORT_ASC, 'company' => SORT_DESC, 'age' => SORT_ASC], $query->orderBy);
+
+        $query->orderBy(new Expression('CASE WHEN x = y THEN 0 ELSE 1 END DESC'));
```
this will internally be cased to a string and not fail as it does not contain a `,` if you add an `Expression` that contains a comma, it should fail.

anestor-sugar replied:
I tried to run following test on my local machine and it passed successfully
```
$query->orderBy(new Expression('CASE WHEN x = y THEN 0 ELSE 1 END DESC, x ASC'));
$this->assertEquals(['CASE WHEN x = y THEN 0 ELSE 1 END' => SORT_DESC, 'x' => SORT_ASC], $query->orderBy);
```

andrewnester replied:
I tried to run following test on my local machine and it passed successfully
```
$query->orderBy(new Expression('CASE WHEN x = y THEN 0 ELSE 1 END DESC, x ASC'));
$this->assertEquals(['CASE WHEN x = y THEN 0 ELSE 1 END' => SORT_DESC, 'x' => SORT_ASC], $query->orderBy);
```

I can update my PR to include this test instead previous one

samdark replied:
Yep. Would be great.

andrewnester committed:
Updated expression string for QueryTest to contain multiple expressions #10535

cebe commented:
Even this case works as expected, what does not work is:

```php
$query->orderBy(new Expression('SUBSTR(name, 3, 4) DESC'));
$this->assertEquals(['SUBSTR(name, 3, 4)' => SORT_DESC], $query->orderBy);
```
and
```php
$query->orderBy(new Expression('SUBSTR(name, 3, 4) DESC, x ASC'));
$this->assertEquals(['SUBSTR(name, 3, 4)' => SORT_DESC, 'x' => SORT_ASC], $query->orderBy);
```

andrewnester committed:
Updated regexp for normalizeOrderBy not to process aggregator functons #10535

andrewnester commented:
@cebe You are absolutely right! I've added new commit with changes to regular expression used in `normalizeOrderBy` and updated tests. Now aggregate functions in `ORDER BY` clause are processed fine

samdark commented on framework/db/QueryTrait.php:
```diff
@@ -341,7 +341,7 @@ protected function normalizeOrderBy($columns)
             return $columns;
         } else {
-            $columns = preg_split('/\s*,\s*/', trim($columns), -1, PREG_SPLIT_NO_EMPTY);
+            $columns = preg_split('/\s*,(?![^\(]*\))\s*/', trim($columns), -1, PREG_SPLIT_NO_EMPTY);
```
Why do we need grouping here?

andrewnester replied:
Actually, it's not grouping. it's `if-else` syntax for regular expression
http://www.regular-expressions.info/conditional.html

samdark replied:
Ah, that's negative lookahead. Haven't used it for a while. Verified that solution works well.

samdark commented:
This one works well. We can merge it either for 2.0.7 or 2.0.8. @cebe ?

cebe commented:
When an Expression object is passed I'd prefer it to be kept as is (not casting to string) until it reaches query builder. This would be constistent with how Expressions are handled when passed in an array:

https://github.com/yiisoft/yii2/blob/c79ebf49f4e9ae28a570ac44e0d6b3015b00dc74/framework/db/QueryBuilder.php#L801

A proper fix would be to check for `$columns instanceof Expression` and then `return [$columns];`

andrewnester committed:
Expression objects for orderBy should be handled in QueryBuilder #10535

andrewnester commented:
@cebe I've just added new commit with fix you offered

cebe commented:
Merged with some adjustments. Thank you!
